pyomo/core/base/units_container.py

```diff
@@ -0,0 +1,1335 @@
+        node_type = type(node)
+        if node_type in self.node_type_method_map:
+            node_func = self.node_type_method_map[node_type]
+            if node_func is not None:
+                pyomo_unit, pint_unit = node_func(self, node, data)
+                if pint_unit is not None and \
+                    self._pint_unit_equivalent_to_dimensionless(pint_unit):
+                    # I want to return None instead of dimensionless
+                    # but radians are also equivalent to dimensionless
+                    # For now, this test works, but we need to find a better approach
+                    teststr = '{:!~s}'.format(pint_unit)
+                    if teststr == '':
+                        return (None, None)
+
+                return (pyomo_unit, pint_unit)
+
+        raise TypeError('An unhandled expression node type: {} was encountered while retrieving the'
+                        ' units of expression'.format(str(node_type), str(node)))
+
+
+class PyomoUnitsContainer(object):
+    """Class that is used to create and contain units in Pyomo.
+
+    This is the class that is used to create, contain, and interact with units in Pyomo.
+    The module (:mod:`pyomo.core.base.units_container`) also contains a module attribute
+    called `units` that is a singleton instance of a PyomoUnitsContainer. This singleton should be
+    used instead of creating your own instance of a :py:class:`PyomoUnitsContainer`.
+    For an overview of the usage of this class, see the module documentation
+    (:mod:`pyomo.core.base.units_container`)
+
+    This class is based on the "pint" module. Documentation for available units can be found
+    at the following url: https://github.com/hgrecco/pint/blob/master/pint/default_en.txt
+
+    Note: Pre-defined units can be accessed through attributes on the PyomoUnitsContainer
+    class; however, these attributes are created dynamically through the __getattr__ method,
+    and are not present on the class until they are requested.
+    """
+    def __init__(self):
+        """Create a PyomoUnitsContainer instance. """
+        # Developers: Do not interact with this attribute directly, but instead
+        # access through the property _pint_registry since that is where the import
+        # of the 'pint' module is checked
+        self.__pint_registry = None
+
+    @property
+    def _pint_registry(self):
+        """ Return the pint.UnitsRegistry instance corresponding to this container. """
+        if pint_module is None:
+            # pint was not imported for some reason
+            raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
+                              " the package 'pint', but this package could not be imported."
+                              " Please make sure you have 'pint' installed.")
+
+        if self.__pint_registry is None:
+            self.__pint_registry = pint_module.UnitRegistry()
+
+        return self.__pint_registry
```

Comment:
This has two checks against None every time someone gets the _pint_registry, and could be very expensive (plus the overhead for the property call).  What do you think of the following alternative implementation:
  - Just have the PyomoUnitsContainer directly and unconditionally access the module-level pint_module
  - Swap out the global registry if pint is unavailable.
```python
class PyomoUnitsContainer(object):
    def __init__(self):
        if pint_module is None:
            raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
                              " the package 'pint', but this package could not be imported."
                              " Please make sure you have 'pint' installed.")
        self._pint_registry = pint_module.UnitRegistry()
    # ....

class PyomoUnitsContainer_NoPint(object):
    def __getattr__(self, attr):
        raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
                              " the package 'pint', but this package could not be imported."
                              " Please make sure you have 'pint' installed.")
    def __setattr__(self, attr, value):
        raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
                              " the package 'pint', but this package could not be imported."
                              " Please make sure you have 'pint' installed.")

if pint_module is None:
    units = PyomoUnitsContainer_NoPint()
else:
    units = PyomoUnitsContainer()
```

Reply:
See issue #899